### PR TITLE
WIP- #40: Constant term types

### PR DESCRIPTION
**Pull request by Eman-Shaaban**
https://github.com/cassandra-scala/troy/issues/40

### CODE REVIEW

File: cql-ast/src/main/scala/troy/cql/ast/Term.scala [hunk 1]
```diff
@@ -1,8 +1,21 @@
...
 
+import java.sql.Blob
+import java.util.UUID
+
 sealed trait Term
 
-final case class Constant(raw: String) extends Term
+
+sealed trait Constant extends Term
+final case class StringConstant (raw: String) extends Constant
```
Comment (tabdulradi) [inline cql-ast/src/main/scala/troy/cql/ast/Term.scala]: Don't leave a space between class name and parenthesis.

File: cql-ast/src/main/scala/troy/cql/ast/Term.scala [hunk 2]
```diff
@@ -1,8 +1,21 @@
...
+final case class IntegerConstant (raw: Int) extends Constant
+final case class FloatConstant (raw: Float) extends Constant
+final case class BooleanConstant (raw: Boolean) extends Constant
+final case class UuidConstant (raw: UUID) extends Constant
+final case class BlobConstant (raw: Blob) extends Constant
+final case class NullConstant (raw: Null) extends Constant
```
Comment (tabdulradi) [inline cql-ast/src/main/scala/troy/cql/ast/Term.scala]: this can be `case object`, null is a null, no value to capture.

File: cql-ast/src/main/scala/troy/cql/ast/Term.scala [hunk 3]
```diff
@@ -1,8 +1,21 @@
...
+import java.util.UUID
+
 sealed trait Term
 
-final case class Constant(raw: String) extends Term
+
+sealed trait Constant extends Term
+final case class StringConstant (raw: String) extends Constant
+final case class IntegerConstant (raw: Int) extends Constant
+final case class FloatConstant (raw: Float) extends Constant
```
Comment (tabdulradi) [inline cql-ast/src/main/scala/troy/cql/ast/Term.scala]: maybe `raw` can be renamed to `value` or something similar?

File: cql-ast/src/main/scala/troy/cql/ast/Term.scala [hunk 4]
```diff
@@ -1,8 +1,21 @@
...
+final case class BooleanConstant (raw: Boolean) extends Constant
+final case class UuidConstant (raw: UUID) extends Constant
+final case class BlobConstant (raw: Blob) extends Constant
```
Comment (tabdulradi) [inline cql-ast/src/main/scala/troy/cql/ast/Term.scala]: I wouldn't use `java.sql.Blob`, maybe `java.nio.ByteBuffer`, or just a String.

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala [hunk 5]
```diff
@@ -53,7 +53,7 @@ object CqlParser extends JavaTokenParsers
    */
   def constant: Parser[Constant] = {
     import Constants._
-    (string | uuid | number | boolean) ^^ Constant // | hex // TODO
+    (string | uuid | number | boolean) ^^ StringConstant // | hex // TODO
```
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: Now we have a separate class for every type of constant, shouldn't we start using them instead of mapping everything as a string?

File: cql-parser/src/test/scala/troy/cql/parser/dml/SelectStatementParserTest.scala [hunk 6]
```diff
@@ -323,7 +323,7 @@ class SelectStatementParserTest extends FlatSpec with Matchers {
     relations.size shouldBe 1
     relations(0).asInstanceOf[Relation.Simple].columnName shouldBe "weight"
     relations(0).asInstanceOf[Relation.Simple].operator shouldBe Operator.Equals
-    relations(0).asInstanceOf[Relation.Simple].term.asInstanceOf[FloatConstant].value shouldBe "50.4"
+    relations(0).asInstanceOf[Relation.Simple].term.asInstanceOf[FloatConstant].value shouldBe 50.4
```
Comment (Eman-Shaaban) [inline cql-parser/src/test/scala/troy/cql/parser/dml/SelectStatementParserTest.scala]: This test fails !

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala [hunk 7]
```diff
@@ -53,7 +55,13 @@ object CqlParser extends JavaTokenParsers
    */
   def constant: Parser[Constant] = {
     import Constants._
-    (string | uuid | number | boolean) ^^ Constant // | hex // TODO
+    def str = (string | uuid) ^^ StringConstant
+    def int = integer ^^ { s => new IntegerConstant(s.toInt) }
+    def floatNum = float ^^ { s => new FloatConstant(s.toFloat) }
+    //def uuid = uuid ^^ { s => new UuidConstant(new UUID (UUID.fromString(s))) }
```
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: What is the problem with this line?

Comment (Eman-Shaaban) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: The input should be String but uuid is an Expression 
"[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}".r

File: cql-parser/src/test/scala/troy/cql/parser/dml/SelectStatementParserTest.scala (hunk 6 above)
Comment (tabdulradi) [inline cql-parser/src/test/scala/troy/cql/parser/dml/SelectStatementParserTest.scala]: This because of the nature of floats being not very accurate. You should allow some variance, search for methods that accepts deltas or ranges.

Comment (tabdulradi) [inline cql-parser/src/test/scala/troy/cql/parser/dml/SelectStatementParserTest.scala]: This because of the nature of floats being not very accurate. You should allow some variance, search for methods that accepts deltas or ranges.

Comment (Eman-Shaaban) [inline cql-parser/src/test/scala/troy/cql/parser/dml/SelectStatementParserTest.scala]: I searched in the scalatest repo and there is many issues about Float numbers, I am not sure if this is a bug or something else.

Comment (tabdulradi) [inline cql-parser/src/test/scala/troy/cql/parser/dml/SelectStatementParserTest.scala]: It is not a bug, this is the nature of the floats. Try executing `0.1.toFloat + 0.2` in scala console. 
Read this for more info https://www.quora.com/How-is-a-float-or-double-value-stored-in-memory-in-java

Comment (tabdulradi) [inline cql-parser/src/test/scala/troy/cql/parser/dml/SelectStatementParserTest.scala]: I also I believe writing the number as `50.4f` should solve the problem.

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala (hunk 7 above)
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: There is some magic conversion happening here, `uuid` is a  `Regex`, but you can call any parser methods on it, as if it is a `Parser[String]` (IntellJ should underline the `^^` operator to denote that there is implicit conversation happened). 
However there 2 issues in the code:
1. `def uuid = uuid`, you choose exactly the same name! so you masked the original uuid. 
2. No need for `new UUID` part, `UUID.fromString` return the UUID instance you want.

**Comment by tabdulradi** [comment]
Good job so far! I know Parser combinators has a high learning curve, a lot of weird symbols and magic implicit conversions, but once you get the grip of it, it will be straight forward to implement any grammar.

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala [hunk 8]
```diff
@@ -53,7 +55,13 @@ object CqlParser extends JavaTokenParsers
...
+    def bool = boolean ^^ { s => new BooleanConstant(s.toBoolean) }
+
+    str | int | floatNum | bool
```
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: Try swapping `int` and `floatNum`
This should fix the failing test case related to FloatConstant. This is due to the fact that order matters, as int matches on the float numbers till it reaches the dot, so it gives the error without backtracking

File: cql-parser/src/test/scala/troy/cql/parser/dml/SelectStatementParserTest.scala (hunk 6 above)
Comment (tabdulradi) [inline cql-parser/src/test/scala/troy/cql/parser/dml/SelectStatementParserTest.scala]: Oh didn't notice that the whole query doesn't even parse. Check my previous comment related to the order of matchers.

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala [hunk 9]
```diff
@@ -53,7 +55,20 @@ object CqlParser extends JavaTokenParsers
    */
   def constant: Parser[Constant] = {
     import Constants._
-    (string | uuid | number | boolean) ^^ Constant // | hex // TODO
```
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: Let's restore this line, and move the mapping logic inside to each matcher separately. 
The only problem you will face is that Constants.string is used in different places as Parser[String] not Parser[StringConstant], so you may define another version outside `Constants` for other usages.

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala [hunk 10]
```diff
@@ -53,7 +55,14 @@ object CqlParser extends JavaTokenParsers
...
+    def str = string ^^ StringConstant
+    def floatNum = float ^^ { s => new FloatConstant(s.toFloat) }
+    def int = integer ^^ { s => new IntegerConstant(s.toInt) }
+    def uuidNum = uuid ^^ { s => new UuidConstant(UUID.fromString(s)) }
+    def bool = boolean ^^ { s => new BooleanConstant(s.toBoolean) }
+    def nullconst = "".r ^^^ NullConstant
```
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: This is not correct

File: cql-ast/src/main/scala/troy/cql/ast/Term.scala [hunk 11]
```diff
@@ -1,8 +1,18 @@
 package troy.cql.ast
 
+import java.util.UUID
+
 sealed trait Term
 
-final case class Constant(raw: String) extends Term
+sealed trait Constant extends Term
+final case class IntegerConstant(value: Int) extends Constant
+final case class FloatConstant(value: Float) extends Constant
```
Comment (tabdulradi) [inline cql-ast/src/main/scala/troy/cql/ast/Term.scala]:
```
float    ::=  re('-?[0-9]+(\.[0-9]*)?([eE][+-]?[0-9+])?') | NAN | INFINITY
```
Seems that float itself needs to be a sealed trait

File: cql-parser/src/test/scala/troy/cql/parser/ddl/AlterTableParserTest.scala [hunk 12]
```diff
@@ -109,10 +109,10 @@ class AlterTableParserTest extends FlatSpec with Matchers {
 
     val option1 = alertTableInstruction1.options(0).asInstanceOf[ConstantOption]
     option1.basicIdentifier shouldBe "comment"
-    option1.constant shouldBe Constant("A most excellent and useful table")
+    option1.constant shouldBe StringConstant("A most excellent and useful table")
 
     val option2 = alertTableInstruction1.options(1).asInstanceOf[ConstantOption]
     option2.basicIdentifier shouldBe "read_repair_chance"
-    option2.constant shouldBe Constant("0.2")
+    option2.constant shouldBe StringConstant("0.2")
```
Comment (tabdulradi) [inline cql-parser/src/test/scala/troy/cql/parser/ddl/AlterTableParserTest.scala]: Shouldn't this be FloatConstant?

File: cql-parser/src/test/scala/troy/cql/parser/dml/DeleteStatementParserTest.scala [hunk 13]
```diff
@@ -21,12 +22,12 @@ class DeleteStatementParserTest extends FlatSpec with Matchers {
...
     simpleRelation.columnName shouldBe "movie"
     simpleRelation.operator shouldBe Operator.Equals
-    simpleRelation.term.asInstanceOf[Constant].raw shouldBe "Serenity"
+    simpleRelation.term.asInstanceOf[StringConstant].value shouldBe "Serenity"
 
   }
 
   it should "parse simple delete statement with UUID value" in {
-    val statement = parseQuery("DELETE FROM Users WHERE userid = 01234567-0123-0123-0123-0123456789ab;")
+    val statement = parseQuery("DELETE FROM Users WHERE userid = B70DE1D0-9908-4AE3-BE34-5573E5B09F14;")
```
Comment (tabdulradi) [inline cql-parser/src/test/scala/troy/cql/parser/dml/DeleteStatementParserTest.scala]: Never ever skew an existing testcase so that your code passes!
The point of existing testcases is to warn you if your changes are doing something wrong or broke an existing feature. 

The value you removed was a valid UUID and should work, you made a mistake in the parser caused this case to fail. You should restore the old value and figure out what went wrong.

File: cql-parser/src/test/scala/troy/cql/parser/dml/SelectStatementParserTest.scala [hunk 14]
```diff
@@ -265,12 +268,12 @@ class SelectStatementParserTest extends FlatSpec with Matchers {
...
     relations(0).asInstanceOf[Relation.Simple].columnName shouldBe "userid"
     relations(0).asInstanceOf[Relation.Simple].operator shouldBe Operator.Equals
-    relations(0).asInstanceOf[Relation.Simple].term.asInstanceOf[Constant].raw shouldBe "199"
+    relations(0).asInstanceOf[Relation.Simple].term.asInstanceOf[IntegerConstant].value shouldBe 199
   }
 
   it should "parse select statements with simple where clause with UUID term" in {
     val statement = parseQuery(
-      "SELECT JSON name, occupation FROM users WHERE userid = 01234567-0123-0123-0123-0123456789ab;"
+      "SELECT JSON name, occupation FROM users WHERE userid = B70DE1D0-9908-4AE3-BE34-5573E5B09F14;"
```
Comment (tabdulradi) [inline cql-parser/src/test/scala/troy/cql/parser/dml/SelectStatementParserTest.scala]: Same as previous comment

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala [hunk 15]
```diff
@@ -72,7 +81,7 @@ object CqlParser extends JavaTokenParsers
 
     def integer = wholeNumber
 
-    def float = floatingPointNumber
+    def float = "[+-]?[0-9]*((\\.[0-9]+([eE][+-]?[0-9]+)?[fF]?)|([fF])|([eE][+\u200C\u200B-]?[0-9]+))\\b".r
```
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: Based on [Codacy's code review](https://www.codacy.com/app/tamer/troy/pullRequest?prid=424169&bid=3921794), this line has non-ascii character, I don't know which, but we need to fix it.

Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: Scala has triple quotes, which doesn't need those extra escapes.

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala [hunk 16]
```diff
@@ -72,7 +81,7 @@ object CqlParser extends JavaTokenParsers
 
     def integer = wholeNumber
 
-    def float = floatingPointNumber
+    def float = """[+-]?[0-9]*((\.[0-9]+([eE][+-]?[0-9]+)?[fF]?)|([fF])|([eE][+\u200C\u200B-]?[0-9]+))\b""".r
```
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: The float regex specified by Datastax docs is `-?[0-9]+(\.[0-9]*)?([eE][+-]?[0-9+])?`, can we use it instead?

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala [hunk 17]
```diff
@@ -53,7 +55,18 @@ object CqlParser extends JavaTokenParsers
    */
   def constant: Parser[Constant] = {
     import Constants._
-    (string | uuid | number | boolean) ^^ Constant // | hex // TODO
+    def str = string ^^ StringConstant
```
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: Most of parsers defined in Constants object are only used here. Maybe you can cut/paste them here instead of having dummy variable names like "string" and "str".

File: cql-ast/src/main/scala/troy/cql/ast/Term.scala [hunk 18]
```diff
@@ -1,8 +1,23 @@
...
+import java.util.UUID
+
 sealed trait Term
 
-final case class Constant(raw: String) extends Term
+sealed trait Constant extends Term
+
+sealed trait FloatNum extends Constant
+final case class FloatConstant(value: Float) extends FloatNum
+case object Nan extends FloatNum
```
Comment (tabdulradi) [inline cql-ast/src/main/scala/troy/cql/ast/Term.scala]: can we rename this to NaN?

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala [hunk 19]
```diff
@@ -53,7 +55,25 @@ object CqlParser extends JavaTokenParsers
...
   def constant: Parser[Constant] = {
     import Constants._
-    (string | uuid | number | boolean) ^^ Constant // | hex // TODO
+
+    def float = """[+-]?[0-9]*((\.[0-9]+([eE][+-]?[0-9]+)?[fF]?)|([fF])|([eE][+‌​-]?[0-9]+))\b""".r
+    def hex = "[0-9a-fA-F]".r
+    def uuid = s"$hex{8}-$hex{4}-$hex{4}-$hex{4}-$hex{12}".r
+    def blob = s"0(x|X)$hex+".r
+
+    def floatNum = float ^^ { s => new FloatConstant(s.toFloat) }
```
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: Can we inline float, uuid, and blob? like:
```
def float = """[+-]?[0-9]*((\.[0-9]+([eE][+-]?[0-9]+)?[fF]?)|([fF])|([eE][+‌​-]?[0-9]+))\b""".r ^^ { s => 
  new FloatConstant(s.toFloat) 
}
```

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala [hunk 20]
```diff
@@ -53,7 +55,25 @@ object CqlParser extends JavaTokenParsers
...
+    def nan = "NaN".r ^^^ Nan
```
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: Can you add a test case that ensures "NAN", and "Nan" won't parse?
I can imagine myself "fixing" those to be case insensitive.

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala [hunk 21]
```diff
@@ -53,7 +55,25 @@ object CqlParser extends JavaTokenParsers
...
+    def infinity = "Infinity".r ^^^ Infinity
```
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: Same as above

File: cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala (hunk 9 above)
Comment (tabdulradi) [inline cql-parser/src/main/scala/troy/cql/parser/CqlParser.scala]: Also a comment in the tests to describe that NaN is case sensitive while Null is not.

**Comment by tabdulradi** [comment]
LGTM in general. Only a few nitpicks then we are good to merge.

**Comment by tabdulradi** [comment]
Thanks very much for your contribution.
